src/sentry/search/utils.py

```diff
@@ -240,58 +242,95 @@ def tokenize_query(query):
-                    break
-                value = '%s %s' % (value, nvalue)
-
-            if value[-1] == '"':
-                value = value[1:-1]
-            else:
-                value = value[1:]
-        results[key].append(value)
-    return dict(results)
+    result = defaultdict(list)
+    query_params = {'tags': [], 'query': []}
+    tokens = split_query_into_tokens(query)
+    for token in tokens:
+        state = 'query'
+        for idx, char in enumerate(token):
+            if idx == 0 and char in ['"', "'"]:
+                break
+            if char == ':':
+                if token[idx + 1] == ':':
+                    state = 'query'
+                elif token[idx + 1] == ' ':
+                    state = 'query'
+                else:
+                    state = 'tags'
+                break
+        query_params[state].append(token)
+
+    result['query'] = map(format_query, query_params['query'])
+    for tag in query_params['tags']:
+        key, value = format_tag(tag)
+        result[key].append(value)
+    return result
+
+
+def format_tag(tag):
+    '''
+    Splits tags on ':' and removes enclosing quotes if present and returns
+    returns both sides of the split as strings
+
+    Example:
+    >>> format_tag('user:foo')
+    'user', 'foo'
+    >>>format_tag('user:"foo bar"'')
+    'user', 'foo bar'
+    '''
+    idx = tag.index(':')
+    key = tag[:idx].strip('"')
+    value = tag[idx + 1:].strip('"')
+    return key, value
+
+
+def format_query(query):
+    '''
+    Strips enclosing quotes from queries if present.
+
+    Example:
+    >>> format_query('"user:foo bar"')
+    'user:foo bar'
+    '''
+    return query.strip('"')
+
+
+def split_query_into_tokens(query):
+    '''
+    Splits query string into tokens for parsing by 'tokenize_query'.
+    Returns list of strigs
+    Rules:
+    Split on whitespace
+        Unless
+        - inside enclosing quotes -> 'user:"foo    bar"'
+        - end of last word is a ':' -> 'user:  foo'
+
+    Example:
+    >>> split_query_into_tokens('user:foo user: bar  user"foo bar' foo  bar) =>
+    ['user:foo', 'user: bar', 'user"foo bar"', 'foo',  'bar']
+    '''
+    tokens = []
+    token = ''
+    quote_enclosed = False
+    quote_type = None
+    end_of_prev_word = None
+    for idx, char in enumerate(query):
+        next_char = query[idx + 1] if idx < len(query) - 1 else None
+        token += char
+        if next_char and not re.match('\s', char) and re.match('\s', next_char):
+            end_of_prev_word = char
+        if re.match('\s', char) and not quote_enclosed and end_of_prev_word != ':':
+            if re.search('\w', token):
+                tokens.append(token.strip(' '))
+                token = ''
+        if char in ["'", '"']:
+            if not quote_enclosed or quote_type == char:
+                quote_enclosed = not quote_enclosed
+                if quote_enclosed:
+                    quote_type = char
+    if re.search('\w', token):
```

Comment:
As discussed IRL, we're going to change these checks to be less restrictive as well and work under a a blacklist of special characters instead. Otherwise, this is way too restrictive.